### PR TITLE
Mulit Sector Expansion

### PR DESCRIPTION
**Pull request by YoungFaithful**
This PR includes:
- change from techs.csv -> techs.yml in order to be more flexible with a tree instead of table structure
- multiple carriers (like `electricity`, `hydrogen`...) can be used and an energy balance is calculated for each carrier
- `demand` is a technology to allow different sources of demand of the different carriers
- storage is simplified and split into `storage` that really stores something and `conversion` that converts from one carrier to another
- `:GEN` and `:FLOW` are sparse Variables to reflect that not every technology has a generation or flow in all carriers
- `run_opt` configuration is more general to allow lost-load, lost-emission and emission-limits for different `carrier`s and `impact`s
- sets are defined as `set[dimension][tech_group]` to simplify the different set-groups needed for modular setup
- change of dependencies `ClustForOpt.jl` -> `TimeSeriesClustering.jl`
- issue #20 handled
- `CodeCov` included

### CODE REVIEW

**Comment by holgerteichgraeber** [comment]
These look like really useful additions, I look forward to having a look. Will do so in the coming days. 
In the meantime, do you think it would make sense to update master with the current dev branch, potentially enhanced by a dependency on TimeSeriesClustering instead of ClustForOpt? 
#24 was closed, so it could be useful to have a compatible version in master already.

**Comment by YoungFaithful** [comment]
> These look like really useful additions, I look forward to having a look. Will do so in the coming days.
> In the meantime, do you think it would make sense to update master with the current dev branch, potentially enhanced by a dependency on TimeSeriesClustering instead of ClustForOpt?
> #24 was closed, so it could be useful to have a compatible version in master already.

Thank you. Changing the current dev will cause merge problems, but I addressed the request in a new PR and will take care about merging as soon as we merged the other one.

**Comment by holgerteichgraeber** [comment]
Thank you, that sounds great. I'll revisit this PR over the coming days, it has lots of cool changes and I want to make sure I take the time that I follow how they work.

File: .codecov.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+codecov:
+  notify:
+    require_ci_to_pass: yes
+
+coverage:
+  precision: 2
+  round: down
+  range: "70...100"
```
Comment (holgerteichgraeber) [inline .codecov.yml]: Not sure if I would put a range on that. I see codecoverage more as a tool to check coverage. Little coverage doesn't necessarily mean bad testing, in my opinion it just helps to point out which parts of the code need more testing attention.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+## How to contribute to CapacityExpansion.jl
+Welcome! Thank you for considering to contribute to `CapacityExpansion.jl`. If you have a comment, question, feature request, or bug report, please open a new [issue](https://github.com/YoungFaithful/CapacityExpansion.jl/issues).
```
Comment (holgerteichgraeber) [inline CONTRIBUTING.md:2]: Maybe here or in README: say that master branch is the stable version and dev branch is the development version (we follow the GitFlow)

File: README.md [hunk 3]
```diff
@@ -6,12 +6,17 @@
 
 [CapacityExpansion](https://github.com/YoungFaithful/CapacityExpansion.jl) is a [julia](https://www.juliaopt.com) implementation of a input-data-scaling capacity expansion modeling framework.
 
+The package has three main purposes:
+1) Provide a simple process to integrate (clustered) time-series input data, geographical input data, cost input data, and technology input data.
```
Comment (holgerteichgraeber) [inline README.md]: 0) Provide an extensible, simple-to-use generation and transmission capacity extension model that allows to address a diverse set of research questions in the area of energy systems planning

File: data/CA_1/nodes.csv [hunk 4]
```diff
@@ -1,3 +1,3 @@
-node,region,infrastruct,lat,lon,|,pv,wind,coal,oil,gas,bat_e,bat_in,bat_out,h2_e,h2_in,h2_out,trans
-california,CA,ex,36.782142,-119.415653,|,9857,5741,101,539,45321,115,98,98,0,0,0,0
-california,CA,lim,36.782142,-119.415653,|,1000000,1000000,100000,100000,100000,1000000,1000000,100000,10000000000,1000000,1000000,0
+node,region,infrastruct,lat,lon,|,pv,wind,coal,oil,gas,bat_e,bat_in,bat_out,h2_e,h2_in,h2_out,trans,demand
```
Comment (holgerteichgraeber) [inline data/CA_1/nodes.csv:1]: I see that demand is a technology now. How do we make sure that that demand is always used, even if we don't set any pre-existing technologies (greenfield)?

File: data/CA_1/techs.yml [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        name: Oil Plant
+        tech_group: dispatchable_generation
+        plant_lifetime: 30
+        financial_lifetime: 30
+        input:
+            fuel: oil
+        output:
+            carrier: electricity
+        constraints:
+            max_gradient: 3 #not yet funcional
```
Comment (holgerteichgraeber) [inline data/CA_1/techs.yml:79]: Maybe add comment with unit? - potentially in general in the settings yaml files.

File: docs/src/csv_structure.md [hunk 6]
```diff
@@ -50,17 +50,21 @@ The package provides data for:
...
-    - `[storage-name]_e`: The energy part of the storage device [MWh]
-    - `[storage-name]_in`: The power part for charging the storage device [MW]
-    - `[storage-name]_out`: The power part for discharging the storage device [MW]
-    If e.g. in a lithium-ion battery the `storage_in` should be the same as `storage_out`, just set the `cap` costs in `costs.csv` of either `storage_in` or `storage_out` to zero. This will add a constraint to bind their capacities.
 
-|`tech`|`categ`|`sector`|`fuel`|`eff`|`max_gradient`|`time_series`|`lifetime`|`financial_lifetime`|`discount_rate`|
-|-------|--------|------|-----|--------|-------|-----------------|------------|----------|--------------------|
-|[tech...]| function handeling those |`el` for electricity|`none` or fuel dependency|efficiency |max gradient of this technology| `none` or time-series name of this tech|lifetime of an installed cap|time in which you have to pay back your loan| `discount_rate`|
-|... |... |... |... |... |... |... |... |... |... |
+`techs.yml` needs to have the following structure:
```
Comment (holgerteichgraeber) [inline docs/src/csv_structure.md:69]: Mention which ones of these define sets in the optimization problem

File: docs/src/csv_structure.md [hunk 7]
```diff
@@ -50,17 +50,21 @@ The package provides data for:
...
+- `tech_groups`: defines parent groups for techs or other tech_groups
```
Comment (holgerteichgraeber) [inline docs/src/csv_structure.md:70]: what is a parent group and why is it needed?

File: docs/src/csv_structure.md [hunk 8]
```diff
@@ -50,17 +50,21 @@ The package provides data for:
...
+- `techs`: defines the technologies, the elements are used to declare the dimension `tech`
+The information of the single techs is combined with the information provided within parent tech_groups. One technology can have multiple tech_groups, if the parant has a parant. The combined information must contain:
+- `name`: A detailed name of the technology
```
Comment (holgerteichgraeber) [inline docs/src/csv_structure.md]:
```suggestion
- `name`: A detailed description of the technology
```

File: docs/src/csv_structure.md [hunk 9]
```diff
@@ -50,17 +50,21 @@ The package provides data for:
...
+- `tech_group`: a technology-group that the technology belongs to. Groups can be: `all`, `demand`, `generation`, `dispatchable_generation`, `non_dispatchable_generation`, `storage`, `conversion`, `transmission`
+- `plant_lifetime`: the lifetime of this technologies plant [a]
```
Comment (holgerteichgraeber) [inline docs/src/csv_structure.md]:
```suggestion
- `plant_lifetime`: the physical expected lifetime of this technologies plant [a]
```

File: docs/src/index.md [hunk 10]
```diff
@@ -2,16 +2,21 @@
...
 [![](https://img.shields.io/badge/docs-stable-blue.svg)](https://YoungFaithful.github.io/CapacityExpansion.jl/stable)
 [![](https://img.shields.io/badge/docs-dev-blue.svg)](https://YoungFaithful.github.io/CapacityExpansion.jl/dev)
-[![Build Status](https://travis-ci.com/holgerteichgraeber/ClustForOpt.jl.svg?token=HRFemjSxM1NBCsbHGNDG&branch=master)](https://travis-ci.com/YoungFaithful/CapacityExpansion.jl)
+[![Build Status](https://travis-ci.com/holgerteichgraeber/TimeSeriesClustering.jl.svg?token=HRFemjSxM1NBCsbHGNDG&branch=master)](https://travis-ci.com/YoungFaithful/CapacityExpansion.jl)
 
-[CapacityExpansion](https://github.com/YoungFaithful/CapacityExpansion.jl) is a [julia](https://www.juliaopt.com) implementation of a input-data-scaling capacity expansion modeling framework.
+[CapacityExpansion](https://github.com/YoungFaithful/CapacityExpansion.jl) is a [julia](https://www.juliaopt.com) implementation of a input-data-scaling, multi-sector capacity expansion modeling framework.
+
+The package has three main purposes:
+1) Provide a simple process to integrate (clustered) time-series input data, geographical input data, cost input data, and technology input data.
```
Comment (holgerteichgraeber) [inline docs/src/index.md]: see other comment in README

File: docs/src/index.md [hunk 11]
```diff
@@ -23,8 +28,7 @@
 |Time resolution 	    | hourly                                                          									 |
 |Network coverage 	  | transmission, DCOPF load flow                                   								   |
 
-
-The package uses [ClustForOpt](https://github.com/holgerteichgraeber/ClustForOpt.jl) as a basis for it's time-series aggregation.
+The package uses [TimeSeriesClustering](https://github.com/holgerteichgraeber/TimeSeriesClustering.jl) as a basis for it's time-series aggregation.
```
Comment (holgerteichgraeber) [inline docs/src/index.md]:
```suggestion
The package uses [TimeSeriesClustering](https://github.com/holgerteichgraeber/TimeSeriesClustering.jl) as a basis for its time-series aggregation.
```

File: docs/src/opt_cep.md [hunk 12]
```diff
@@ -12,41 +14,50 @@ The model is designed to minimize social costs by minimizing the following objec
...
 | impact           | impact categories like EUR or USD, CO 2 − eq., ...                    |
 | account          | fixed costs for installation and yearly expenses, variable costs      |
 | infrastruct      | infrastructure status being either new or existing                    |
-| sector           | energy sector like electricity                                        |
 | time K           | numeration of the representative periods                              |
-| time T           | numeration of the time intervals within a period                      |
-| time T e         | numeration of the time steps within a period                          |
-| time I           | numeration of the time invervals of the full input data periods       |
-| time I e         | numeration of the time steps of the full input data periods           |
+| time T period    | numeration of the time intervals within a period                      |
```
Comment (holgerteichgraeber) [inline docs/src/opt_cep.md:35]: I do not understand the difference between time point and time interval. Are we talking about T time steps of length `delta t` within a period?

File: docs/src/opt_cep.md [hunk 13]
```diff
@@ -55,25 +66,35 @@ An overview of the variables used in the CEP is provided in the table:
...
+| allowing conversion (necessary for storage)        | -                | `conversion`            | `true` or `false`                               | ::Bool         | false         |
+| allowing demand                                    | -                | `demand`            | `true` or `false`                               | ::Bool         | true         |
+| allowing dispatchable generation                   | -                | `dispatchable_generation`            | `true` or `false`                               | ::Bool         | false         |
+| allowing non dispatchable generation               | -                | `non_dispatchable_generation`            | `true` or `false`                               | ::Bool         | true         |
+| allowing transmission                              | -                | `transmission`            | `true` or `false`                               | ::Bool         | false         |
+| fix. var and CEO to dispatch problem               | -                | `fixed_design_variables`  | design variables from design run or nothing | ::OptVariables | nothing       |
```
Comment (holgerteichgraeber) [inline docs/src/opt_cep.md]: CEO?

File: docs/src/opt_cep.md [hunk 14]
```diff
@@ -84,6 +105,15 @@ using Clp
 optimizer=Clp.Optimizer
 ```
 
+## Solver Configuration
+Depending on the Solver different solver configurations are possible. The information is always provided as `Dict{Symbol,Any}`. The keys of the dictionary are the parameters and the values of the dictionary are the values passed to the solver.
+
+For example the `Gurobi` solver can be configured to have no OutputFlag and run on two threads (per julia thread) the following way:
+```julia
+optimizer_config=Dict{Symbol,Any}(:OutputFlag => 0, :Threads => 2)
```
Comment (holgerteichgraeber) [inline docs/src/opt_cep.md:113]: Just out of curiosity: Have you found that it becomes faster if you run it on multiple threads (but still one core)? Or did you also increase the number of cores?

File: docs/src/opt_cep.md [hunk 15]
```diff
@@ -84,6 +105,15 @@ using Clp
...
+```
+Further information on possible keys for Gurobi can be found at [Gurobi parameter description](https://www.gurobi.com/documentation/8.1/refman/parameter_descriptions.html).
```
Comment (holgerteichgraeber) [inline docs/src/opt_cep.md:115]: Thanks for adding this, great!

File: examples/workflow_example_cep.jl [hunk 16]
```diff
@@ -20,41 +20,41 @@ ts_seg_data = run_clust(ts_input_data;method="kmeans",representation="centroid",
 optimizer=Clp.Optimizer
 
 # tweak the CO2 level
-co2_result = run_opt(ts_clust_data.clust_data,cep_data,optimizer;co2_limit=50) #generally values between 1250 and 10 are interesting
+co2_result = run_opt(ts_clust_data.clust_data,cep_data,optimizer;limit_emission=Dict{String,Number}("CO2/electricity"=>50)) #generally values between 1250 and 10 are interesting
```
Comment (holgerteichgraeber) [inline examples/workflow_example_cep.jl]: What does 'CO2/electricity' mean? How about `kg_CO2/kWh` or `CO2:kg/kWh`? I feel having units somewhere could make things helpful. Similar for lost load.

File: docs/src/opt_cep.md [hunk 17]
```diff
@@ -55,25 +66,35 @@ An overview of the variables used in the CEP is provided in the table:
...
-| allowing lost emission (necessary for dispatch)                    | price/kg-CO2e | `lost_CO2_emission_cost`  | >700                                        | ::Number       | Inf           |
+| enforce an emission-limit                          | kg-impact/MWh-carrier | `limit_emission`               | Dict{String,Number}(impact/carrier=>value)                                      | ::Dict{String,Number}       | Dict{String,Number}()           |
+| including existing infrastructure (no extra costs) and limit infrastructure   | -                | `infrastructure`| Dict{String,Array}("existing"=>[tech-groups...], "limit"=>[tech-groups...])                             | ::Dict{String,Array}       | Dict{String,Array}("existing"=>["demand"])         |
+| type of storage implementation                     | -                | `storage_type`                 | "none", "simple" or "seasonal"              | ::String       | "none"        |
+| allowing conversion (necessary for storage)        | -                | `conversion`            | `true` or `false`                               | ::Bool         | false         |
```
Comment (holgerteichgraeber) [inline docs/src/opt_cep.md:74]: What is conversion and why is it necessary? If we can avoid having settings that depend on each other, that would be great for the new user (e.g. storage and conversion).

File: src/optim_problems/opt.jl [hunk 18]
```diff
@@ -0,0 +1,718 @@
...
+function setup_opt_cost_var!(cep::OptModelCEP,
+                            ts_data::ClustData,
+                            opt_data::OptDataCEP,
+                            scale::Dict{Symbol,Int},
+                            tech_group::String;
+                            dir_main_carrier::Symbol=:input,
+                            sign_generation::Number=1)
+    ## DATA ##
+    set=cep.set
+    #`costs::OptVariable`: costs[tech,node,year,account,impact] - annulized costs [USD in USD/MW_el, CO2 in kg-CO₂-eq./MW_el]`
```
Comment (holgerteichgraeber) [inline src/optim_problems/opt.jl:133]: I like that you have a comment above each line explaining what it does, great!

File: src/utils/datastructs.jl [hunk 19]
```diff
@@ -3,22 +3,31 @@ abstract type OptData <: InputData end
 
 """
      OptModelCEP
--model::JuMP.Model
--info::Array{String}
--set::Dict{String,Array}
+The essential elements of a CapacityExpansionProblem are organized as a `OptModelCEP` struct:
+- `model`::JuMP.Model - contains the actual JuMP-Model
+- `info`::Array{String} - contains the information about the model setup in form of multiple lines of equations
+- `set`::Dict{String,Array} - contains the sets used by the model
```
Comment (holgerteichgraeber) [inline src/utils/datastructs.jl:9]: the second entry is a dict below

File: src/utils/datastructs.jl [hunk 20]
```diff
@@ -3,22 +3,31 @@ abstract type OptData <: InputData end
...
 """
 struct OptModelCEP
   model::JuMP.Model
   info::Array{String}
-  set::Dict{String,Array}
+  set::Dict{String,Dict{String,Array}}
```
Comment (holgerteichgraeber) [inline src/utils/datastructs.jl:14]: What is the reasoning behind this change?

File: src/utils/load_data.jl [hunk 21]
```diff
@@ -21,61 +21,84 @@ function load_timeseries_data_provided(region::String="GER_1";
                               years::Array{Int,1}=[2016],
                               att::Array{String,1}=Array{String,1}())
     # Check for existance of the region in data
-    region in readdir(normpath(joinpath(@__DIR__,"..","..","data"))) || throw(@error "The region $region is not found. The provided regions are: GER_1: Germany 1 node, GER_18: Germany 18 nodes, CA_1: California 1 node, CA_14: California 14 nodes, TX_1: Texas 1 node")
+    region in readdir(normpath(joinpath(@__DIR__,"..","..","data"))) || throw(error("The region $region is not found. The provided regions are: GER_1: Germany 1 node, GER_18: Germany 18 nodes, CA_1: California 1 node, CA_14: California 14 nodes, TX_1: Texas 1 node"))
```
Comment (holgerteichgraeber) [inline src/utils/load_data.jl:24]: Nice!

File: src/utils/utils.jl [hunk 22]
```diff
@@ -46,78 +10,14 @@ function check_column(df::DataFrame, names_array::Array{Symbol,1})
...
-"""
-    get_cep_variable_set(variable::OptVariable,num_index_set::Int)
-Get the variable set from the specific variable and the `num_index_set` like 1
-"""
-function get_cep_variable_set(variable::OptVariable,
-                              num_index_set::Int
-                              )
-                              @warn "get_cep_variable_set deprecated"
-    return variable.axes[num_index_set]
+#Use getindex to return all rows of the DataFrame fulfilling that in column `col` value `val` is found
```
Comment (holgerteichgraeber) [inline src/utils/utils.jl:13]: What does this do in plain English? (what is col and what is val)

File: src/utils/utils.jl [hunk 23]
```diff
@@ -46,78 +10,14 @@ function check_column(df::DataFrame, names_array::Array{Symbol,1})
...
-    get_cep_variable_set(scenario::Scenario,var_name::String,num_index_set::Int)
-Get the variable set from the specific Scenario by indicating the `var_name` e.g. "COST" and the `num_index_set` like 1
-"""
-function get_cep_variable_set(scenario::Scenario,
-                              var_name::String,
-                              num_index_set::Int
-                              )
-                              @warn "get_cep_variable_set deprecated"
-    return  get_cep_variable_set(scenario.opt_res.variables[var_name], num_index_set)
+#Use getindex to return all rows of the DataFrame fulfilling that in column `col` value `val` is found and the column `colon_ind`
```
Comment (holgerteichgraeber) [inline src/utils/utils.jl:18]: Same here, what does this do in terms of intuition (what is col, what is val)?

Comment (holgerteichgraeber) [reply src/utils/utils.jl:18]: Additionally: Is this referred to in the documentation (if it is a nice function to know)?

File: src/utils/utils.jl [hunk 24]
```diff
@@ -316,28 +221,126 @@ function get_met_cap_limit(cep::OptModelCEP, opt_data::OptDataCEP, variables::Di
...
+        limit_dir[impact][String(carrier)]=limit[join([impact carrier], "/")]
+    end
+    return limit_dir
+end
+
+"""
+    text_limit_emission(limit_emission::Dict{String,Dict{String,Number}})
+Return text with the information of the `limit_emission` as a text
+"""
+function text_limit_emission(limit_emission::Dict{String,Dict})
```
Comment (holgerteichgraeber) [inline src/utils/utils.jl:290]: Nice function! Is it in documenation?

**Review by holgerteichgraeber — CHANGES_REQUESTED** [review]
Great additional features! And also great additional comments and documentation! I look forward to rewrite my iterative solve so that it works with this version. 
Some comments in the code, additionally: 
- Do the old tests still work on the new version (Was it possible to use the old JLD2 file with the new interface)? It would be great to compare results without recomputation of the tests, because there are a lot of changes. 
- On that note: It seems that the jld2 was recalculated in the last step. Is there any way to compare at least a subset of the tests to the old version?

File: data/CA_1/nodes.csv (hunk 4 above)
Comment (YoungFaithful) [reply data/CA_1/nodes.csv:1]: Error no demand

File: src/utils/datastructs.jl (hunk 20 above)
Comment (YoungFaithful) [reply src/utils/datastructs.jl:14]: check for example

**Review by holgerteichgraeber — APPROVED** [review]
Thanks for integrating the suggestions. Was good to talk through it on the phone. Looking great.

**Comment by holgerteichgraeber** [comment]
One question that we did not address yet: 
-  Do the old tests still work on the new version (Was it possible to use the old JLD2 file with the new interface)? It would be great to compare results without recomputation of the tests, because there are a lot of changes.
 -  On that note: It seems that the jld2 was recalculated in the last step. Is there any way to compare at least a subset of the tests to the old version?

**Comment by holgerteichgraeber** [comment]
@YoungFaithful can we merge this?
